File: api/source-status.js
// Vercel Serverless Function — real, live connectivity check for the
// atencionsismo API, backing the "Analista" tab's atencionsismo row. GET-only,
// admin-gated. A snapshot (reportes_meta.json) proves the pipeline ran, not
// that the API is reachable NOW — this endpoint answers that second question
// by re-running the same one-minute probe api/reportados.js already uses
// before its full day-walk (see probeApi there). See design.md ADR-3/ADR-4.
//
// Required env in Vercel (Project Settings → Environment Variables):
//   VISITADOS_API_PASS   password for the api="read" account (same as
//     api/reportados.js). Missing/invalid -> ok:false, NOT a 5xx: the
//     endpoint successfully determined the source is down.
// Optional:
//   VISITADOS_API_USER   default juanp.gzmz@gmail.com

const { verifyFirebaseToken, roleFromClaims } = require('./refresh.js');
const { probeApi } = require('./reportados.js');

const FIREBASE_PROJECT_ID = process.env.FIREBASE_PROJECT_ID || 'sismo-agosto-sgred';
const DEFAULT_USER = 'juanp.gzmz@gmail.com';

async function defaultVerify(idToken) {
  const claims = await verifyFirebaseToken(idToken, FIREBASE_PROJECT_ID);
  return { role: roleFromClaims(claims), email: claims && claims.email };
}

async function defaultProbe() {
  const pass = (process.env.VISITADOS_API_PASS || '').trim();
  if (!pass) {
    const err = new Error('VISITADOS_API_PASS no está configurado en Vercel.');
    err.status = 503;
    throw err;
  }
  const user = (process.env.VISITADOS_API_USER || '').trim() || DEFAULT_USER;
  const auth = Buffer.from(`${user}:${pass}`).toString('base64');
  await probeApi(auth);
}

// Factory over an injectable { verify, probe } so the auth-gate and ok/not-ok
// mapping are testable without the Firebase Admin SDK or a real network call
// (api/source-status.test.js). Vercel's default export uses the real
// dependencies; module.exports.handle is exposed only for the self-check.
function handle({ verify = defaultVerify, probe = defaultProbe } = {}) {
  return async (req, res) => {
    if (req.method !== 'GET') {
      res.setHeader('Allow', 'GET');
      return res.status(405).json({ error: 'Method Not Allowed' });
    }

    const authHeader = req.headers.authorization || '';
    const idToken = authHeader.startsWith('Bearer ') ? authHeader.slice(7) : '';
    if (!idToken) return res.status(401).json({ error: 'Autenticación requerida.' });

    let role;
    try {
      ({ role } = await verify(idToken));
    } catch (err) {
      return res.status(401).json({ error: `Token inválido: ${(err && err.message) || err}` });
    }
    if (role !== 'admin') {
      return res.status(403).json({ error: 'Solo administradores pueden ver el estado de las fuentes.' });
    }

    const checked_at = new Date().toISOString();
    try {
      await probe();
      res.setHeader('Cache-Control', 'private, no-store');
      return res.status(200).json({ ok: true, status: 'conectado', detail: null, checked_at });
    } catch (err) {
      res.setHeader('Cache-Control', 'private, no-store');
      return res.status(200).json({
        ok: false,
        status: 'con errores',
        detail: String((err && err.message) || err),
        checked_at,
      });
    }
  };
}

module.exports = handle();
module.exports.handle = handle;

File: backend/app/routers/source_status.py
"""GET /source-status — admin-only, live connectivity check for the
atencionsismo API (design.md ADR-4; backend-platform spec "Admin-gated
route rejects non-admin" (`/source-status` row)).

Ports `api/source-status.js` verbatim. Backs the "Analista" tab's
atencionsismo status row: a snapshot (`reportes_meta.json`) proves the
pipeline ran at some point in the past, not that the API is reachable NOW —
this route answers that second question by re-running the SAME cheap
one-minute probe `app/services/atencionsismo.py`'s day-walk already runs
before its full range fetch (`atencionsismo.probe_api`, extracted in slice
3).

Deliberately never a 5xx for an upstream failure: a down/misconfigured
atencionsismo API (or a missing `VISITADOS_API_PASS`) is a successfully-
determined FACT (`ok: false`), not a backend error — the legacy handler
always answers `200` on both branches (`api/source-status.js:67,70`), which
this route preserves exactly.

No `Cache-Control` header value is invented: the legacy handler explicitly
sets `private, no-store` on both branches (`api/source-status.js:66,69`),
carried over verbatim below. (NOTE for verify: this is UNLIKE
`api/sticker-status.js`, which sets no `Cache-Control` header at all —
confirmed by reading both files; do not conflate the two.)
"""
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

import httpx
from fastapi import APIRouter, Depends
from fastapi.responses import JSONResponse

from app.auth.deps import require_role
from app.services import atencionsismo

# Never touches Firestore/S3 — VISITADOS_API_PASS is a "plain secret" read
# directly by app/services/atencionsismo.py (design.md ADR-4 table), not
# part of the named-client union.
REQUIRED_CLIENTS: tuple[str, ...] = ()

# Verbatim from api/source-status.js:66,69.
CACHE_CONTROL = "private, no-store"

router = APIRouter()


def _checked_at() -> str:
    """`new Date().toISOString()` shape (millisecond precision, trailing
    `Z`) — Python's default `isoformat()` uses microseconds and a `+00:00`
    offset instead, so this is formatted by hand to match the legacy
    payload byte-for-byte."""
    now = datetime.now(timezone.utc)
    return now.strftime("%Y-%m-%dT%H:%M:%S.") + f"{now.microsecond // 1000:03d}Z"


async def _probe() -> None:
    user, password = atencionsismo.credentials_from_env()
    async with httpx.AsyncClient() as client:
        await atencionsismo.probe_api(client, user, password)


@router.get("/source-status")
async def get_source_status(
    claims: dict[str, Any] = Depends(require_role("admin")),
) -> JSONResponse:
    checked_at = _checked_at()
    try:
        await _probe()
        return JSONResponse(
            {"ok": True, "status": "conectado", "detail": None, "checked_at": checked_at},
            headers={"Cache-Control": CACHE_CONTROL},
        )
    except (
        atencionsismo.ApiCredentialsError,
        atencionsismo.ApiUnavailableError,
        httpx.HTTPError,
    ) as exc:
        return JSONResponse(
            {
                "ok": False,
                "status": "con errores",
                "detail": str(exc),
                "checked_at": checked_at,
            },
            headers={"Cache-Control": CACHE_CONTROL},
        )
